### ISSUE TITLE
sqlsmith: internal error, trailing bytes in encoded value

### ISSUE THREAD

**Issue by irfansharif** [cockroachdb/cockroach#104323]
I saw the following assertion trip up in a [CI run](https://teamcity.cockroachdb.com/viewLog.html?tab=buildLog&logTab=tree&filter=debug&expand=all&buildId=10267972).

```
	test artifacts and logs in: /artifacts/sqlsmith/setup=empty/setting=no-mutations/run_1
(sqlsmith.go:240).func3: error: pq: internal error: 35 trailing bytes in encoded value: [15 15 20 64 0 0 1 144 0 0 4 144 0 0 8 116 101 115 116 106 115 111 110 15 16 0 8 148 245 167 155 5 176 156 27]
stmt:
WITH
	with5045 ("c%qol41326")
		AS (
			SELECT
				*
			FROM
				(
					VALUES
						(🙄func273()),
						((NULL, '{"test": "json"}':::JSONB, (), '1992-03-04 16:39:38.000223+00':::TIMESTAMPTZ)),
						(
							(NULL, '[{"EvG": [true, "baz", false], "MPRRONv": 0.2980103397565371, "foo": {"8rNz5y": true, "baz": null, "kRMYs": false}, "uyRcxposI": "foo"}, {"1rMfbBSOg": {"USyO": true}, "Li5": 1.261090300252655, "a": "7c7QH82", "nsAleYB": "rga3g"}, {"bar": 2.439379404706056}, true, 0.23533439446032228]':::JSONB, (), NULL)
						)
				)
					AS "\\uE1B1t%Ab 21645" ("c%qol41326")
		)
SELECT
	"c%40te_ref1478"."c%qol41326" AS "c͚oL41327"
FROM
	with5045 AS "c%40te_ref1478"
GROUP BY
	"c%40te_ref1478"."c%qol41326"
HAVING
	bool_and(true::BOOL)::BOOL
LIMIT
	95:::INT8;
[20:17:12]	--- FAIL: sqlsmith/setup=empty/setting=no-mutations (413.48s)
```

Jira issue: CRDB-28467

**Comment by mgartner** [comment]
SQL that should reproduce this: https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_RoachtestNightlyGceBazel/10267972:id/sqlsmith/setup%3Dempty/setting%3Dno-mutations/run_1/artifacts.zip!/sqlsmith.log

**Comment by cucaroach** [comment]
I'm probably being paranoid but could be related to: https://github.com/cockroachdb/cockroach/pull/101693

**Comment by DrewKimball** [comment]
Full repro should look like this:
```
CREATE FUNCTION ðŸ™„func273()
	RETURNS RECORD
	STABLE
	LANGUAGE SQL
	AS $funcbody$SELECT (NULL,) AS col26062 FROM (VALUES (2137750727:::OID, 1599972205:::OID, NULL), (0:::OID, 492979399:::OID, 2455666027:::OID), (2751738740:::OID, 4243148485:::OID, 51225265:::OID)) AS """tab13358" (col26059, col26060, col26061) WHERE NULL$funcbody$;

WITH
	with5045 ("c%qol41326")
		AS (
			SELECT
				*
			FROM
				(
					VALUES
						(ðŸ™„func273()),
						((NULL, '{"test": "json"}':::JSONB, (), '1992-03-04 16:39:38.000223+00':::TIMESTAMPTZ)),
						(
							(NULL, '[{"EvG": [true, "baz", false], "MPRRONv": 0.2980103397565371, "foo": {"8rNz5y": true, "baz": null, "kRMYs": false}, "uyRcxposI": "foo"}, {"1rMfbBSOg": {"USyO": true}, "Li5": 1.261090300252655, "a": "7c7QH82", "nsAleYB": "rga3g"}, {"bar": 2.439379404706056}, true, 0.23533439446032228]':::JSONB, (), NULL)
						)
				)
					AS "\\uE1B1t%Ab 21645" ("c%qol41326")
		)
SELECT
	"c%40te_ref1478"."c%qol41326" AS "cÍšoL41327"
FROM
	with5045 AS "c%40te_ref1478"
GROUP BY
	"c%40te_ref1478"."c%qol41326"
HAVING
	bool_and(true::BOOL)::BOOL
LIMIT
	95:::INT8;
```
But running on master, 23.1.2, and 23.1.0 results in a build-time error:
```
ERROR: VALUES types tuple{unknown, jsonb, tuple, timestamptz} and tuple{unknown} cannot be matched
```
The JSON syntax is unimplemented in 22.2. 

I think this was likely introduced by #98162 and fixed by #103078, so we should be good to close it out.

**Comment by irfansharif** [comment]
(Confirmed that the SHA the CI run was from didn't include 8b1d21e28272afe99a8882e2f2ce156f71da7f44.)
